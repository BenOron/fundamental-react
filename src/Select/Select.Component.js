import path from 'path';
import React from 'react';
import { Button, Select, Menu, Popover } from '..';
import { ComponentPage, Example } from '../_playground';


export const SelectComponent = () => {

    return (
        <ComponentPage
            description={`The **Select** component lets the user select one of the different options.
                It is more flexible than the normal Select.`}
            sourceModulePath={path.join(__dirname, './Select')}
            title='Select'>

            <Example
                centered
                title='Default Select'>
                <Select>
                    <Popover
                        body={
                            <Menu>
                                <Menu.List>
                                    <Menu.Item url='#'>Option 1</Menu.Item>
                                    <Menu.Item url='#'>Option 2</Menu.Item>
                                    <Menu.Item url='#'>Option 3</Menu.Item>
                                    <Menu.Item url='#'>Option 4</Menu.Item>
                                </Menu.List>
                            </Menu>
                        }
                        control={<Button className='fd-Select__control'>Select</Button>}
                        id='jhqD0555'
                        noArrow
                        useArrowKeyNavigation />
                </Select>

                <Select>
                    <Popover
                        body={
                            <Menu>
                                <Menu.List>
                                    <Menu.Item url='#'>Option 1</Menu.Item>
                                    <Menu.Item url='#'>Option 2</Menu.Item>
                                    <Menu.Item url='#'>Option 3</Menu.Item>
                                    <Menu.Item url='#'>Option 4</Menu.Item>
                                </Menu.List>
                            </Menu>
                        }
                        control={
                            <Button className='fd-Select__control' compact>
                                Select
                            </Button>
                        }
                        id='jhqD0556'
                        noArrow
                        useArrowKeyNavigation />
                </Select>
            </Example>

            <Example
                centered
                description='It can also include complementary information like an icon.'
                title='Select with Icon'>
                <Select>
                    <Popover
                        body={
                            <Menu>
                                <Menu.List>
                                    <Menu.Item url='#'>Option 1</Menu.Item>
                                    <Menu.Item url='#'>Option 2</Menu.Item>
                                    <Menu.Item url='#'>Option 3</Menu.Item>
                                    <Menu.Item url='#'>Option 4</Menu.Item>
                                </Menu.List>
                            </Menu>
                        }
                        control={
                            <Button className='fd-Select__control' glyph='filter'>
                                Select
                            </Button>
                        }
                        id='jhqD0557'
                        noArrow
                        useArrowKeyNavigation />
                </Select>

                <Select>
                    <Popover
                        body={
                            <Menu>
                                <Menu.List>
                                    <Menu.Item url='#'>Option 1</Menu.Item>
                                    <Menu.Item url='#'>Option 2</Menu.Item>
                                    <Menu.Item url='#'>Option 3</Menu.Item>
                                    <Menu.Item url='#'>Option 4</Menu.Item>
                                </Menu.List>
                            </Menu>
                        }
                        control={
                            <Button className='fd-Select__control' compact
                                glyph='filter'>
                                Select
                            </Button>
                        }
                        id='jhqD0558'
                        noArrow
                        useArrowKeyNavigation />
                </Select>
            </Example>

            <Example
                centered
                title='Toolbar Select'>
                <Select standard>
                    <Popover
                        body={
                            <Menu>
                                <Menu.List>
                                    <Menu.Item url='#'>Option 1</Menu.Item>
                                    <Menu.Item url='#'>Option 2</Menu.Item>
                                    <Menu.Item url='#'>Option 3</Menu.Item>
                                    <Menu.Item url='#'>Option 4</Menu.Item>
                                </Menu.List>
                            </Menu>
                        }
                        control={
                            <Button className='fd-Select__control'>
                                Select
                            </Button>
                        }
                        id='jhqD0559'
                        noArrow
                        useArrowKeyNavigation />
                </Select>

                <Select standard>
                    <Popover
                        body={
                            <Menu>
                                <Menu.List>
                                    <Menu.Item url='#'>Option 1</Menu.Item>
                                    <Menu.Item url='#'>Option 2</Menu.Item>
                                    <Menu.Item url='#'>Option 3</Menu.Item>
                                    <Menu.Item url='#'>Option 4</Menu.Item>
                                </Menu.List>
                            </Menu>
                        }
                        control={
                            <Button className='fd-Select__control' compact>
                                Select
                            </Button>
                        }
                        id='jhqD0560'
                        noArrow
                        useArrowKeyNavigation />
                </Select>
            </Example>

            <Example
                centered
                title='Popover width limited to Select'>
                <Select>
                    <Popover
                        body={
                            <Menu>
                                <Menu.List>
                                    <Menu.Item url='#'>Option 1</Menu.Item>
                                    <Menu.Item url='#'>Option 2</Menu.Item>
                                    <Menu.Item url='#'>Option 3</Menu.Item>
                                    <Menu.Item url='#'>Option 4</Menu.Item>
                                </Menu.List>
                            </Menu>
                        }
                        control={
                            <Button className='fd-Select__control'
                                glyph='navigation-down-arrow'>
                                Open the Select
                            </Button>
                        }
                        id='jhqD0561'
                        noArrow
                        widthSizingType='matchTarget' />
                </Select>
            </Example>

            <Example
                centered
                title='Disabled State'>
                <Select>
                    <Popover
                        body={
                            <Menu>
                                <Menu.List>
                                    <Menu.Item url='#'>Option 1</Menu.Item>
                                    <Menu.Item url='#'>Option 2</Menu.Item>
                                    <Menu.Item url='#'>Option 3</Menu.Item>
                                    <Menu.Item url='#'>Option 4</Menu.Item>
                                </Menu.List>
                            </Menu>
                        }
                        control={
                            <Button className='fd-Select__control' disabled
                                glyph='filter'>
                                Select
                            </Button>
                        }
                        disabled
                        id='jhqD0561'
                        noArrow
                        useArrowKeyNavigation />
                </Select>
            </Example>

        </ComponentPage>
    );
};
