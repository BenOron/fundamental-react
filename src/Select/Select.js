import Button from '../Button/Button';
import classnames from 'classnames';
import Popover from '../Popover/Popover';
import PropTypes from 'prop-types';
import React, { useEffect } from 'react';

const Select = React.forwardRef(({
    children,
    className,
    compact,
    disabled,
    disableStyles,
    inline,
    placeholder,
    ...props
}, ref) => {

    useEffect(() => {
        if (!disableStyles) {
            require('fundamental-styles/dist/select.css');
            require('fundamental-styles/dist/list.css');
        }
    }, []);

    const SelectClasses = classnames(
        'fd-select',
        {
            'fd-select--compact': compact,
            'fd-select--inline': inline
        },
        className
    );

    const SelectControl = (<div {...props} className={SelectClasses}
        ref={ref}>
        <div className='fd-select__control'>
            {placeholder}
            <Button
                className='fd-select__button'
                disabled={disabled}
                glyph='slim-arrow-down'
                option='light'
                ref={ref} />
        </div>
    </div>);

    const popoverBody = () => {
        return React.cloneElement(children, {
            className: 'fd-list--dropdown'
        });
    };

    return (
        <Popover
            body={popoverBody()}
            control={SelectControl}
            noArrow
            placement='bottom-end' />
    );
});

Select.displayName = 'Select';

Select.propTypes = {
    children: PropTypes.node,
    className: PropTypes.string,
    compact: PropTypes.bool,
    disabled: PropTypes.bool,
    disableStyles: PropTypes.bool,
    inline: PropTypes.bool,
    placeholder: PropTypes.string
};


export default Select;
