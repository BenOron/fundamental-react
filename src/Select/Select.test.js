import { Button } from '..';
import Select from './Select';
import Menu from '../Menu/Menu';
import { mount } from 'enzyme';
import Popover from '../Popover/Popover';
import React from 'react';
import renderer from 'react-test-renderer';

describe('<Select />', () => {
    const defaultMenu = (
        <Menu>
            <Menu.List>
                <Menu.Item url='/'>Option 1</Menu.Item>
                <Menu.Item url='/'>Option 2</Menu.Item>
                <Menu.Item url='/'>Option 3</Menu.Item>
                <Menu.Item url='/'>Option 4</Menu.Item>
            </Menu.List>
        </Menu>
    );

    const defaultSelect = (
        <Select>
            <Popover
                body={defaultMenu}
                control={<Button className='fd-Select__control'>Select</Button>}
                noArrow popperProps={{ id: 'fd-default-Select-popover' }} />
        </Select>
    );

    const compactSelect = (
        <Select className='blue'>
            <Popover
                body={defaultMenu}
                control={
                    <Button className='fd-Select__control' compact>
                        Select
                    </Button>
                }
                noArrow popperProps={{ id: 'fd-compact-Select-popover' }} />
        </Select>
    );

    const toolbarSelect = (
        <Select standard>
            <Popover
                body={defaultMenu}
                control={
                    <Button className='fd-Select__control'>
                        Select
                    </Button>
                }
                noArrow popperProps={{ id: 'fd-toolbar-Select-popover' }} />
        </Select>
    );

    const iconSelect = (
        <Select>
            <Popover
                body={defaultMenu}
                control={
                    <Button className='fd-Select__control' glyph='filter'>
                        Select
                    </Button>
                }
                id='jhqD0557'
                noArrow popperProps={{ id: 'fd-icon-Select-popover' }} />
        </Select>
    );

    const disabledSelect = (
        <Select>
            <Popover
                body={defaultMenu}
                control={
                    <Button className='fd-Select__control'
                        disabled
                        glyph='filter'>
                        Select
                    </Button>
                }
                disabled
                id='jhqD0561'
                noArrow popperProps={{ id: 'fd-disable-Select-popover' }} />
        </Select>
    );

    test('create Select component', () => {
        // default Select
        let component = renderer.create(defaultSelect);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // compact Select
        component = renderer.create(compactSelect);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // toolbar Select
        component = renderer.create(toolbarSelect);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // icon Select
        component = renderer.create(iconSelect);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        // disabled Select
        component = renderer.create(disabledSelect);
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    describe('Prop spreading', () => {
        test('should allow props to be spread to the Select component', () => {
            const element = mount(
                <Select data-sample='Sample'>
                    <Popover
                        body={defaultMenu}
                        control={<Button className='fd-Select__control'>Select</Button>}
                        noArrow />
                </Select>
            );

            expect(element.getDOMNode().attributes['data-sample'].value).toBe('Sample');
        });
    });

    test('forwards the ref', () => {
        let ref;
        class Test extends React.Component {
            constructor(props) {
                super(props);
                ref = React.createRef();
            }
            render = () => <Select ref={ref} />;
        }
        mount(<Test />);
        expect(ref.current.tagName).toEqual('DIV');
        expect(ref.current.className).toEqual('fd-Select');
    });
});
