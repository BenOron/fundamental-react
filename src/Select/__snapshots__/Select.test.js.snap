// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> create Select component 1`] = `
<div
  className="fd-popover"
>
  <div
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="fd-popover__control"
    >
      <div
        aria-controls="fd-7WmJd56s"
        aria-expanded={false}
        aria-haspopup={true}
        className="fd-select"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="fd-select__control"
        >
          <button
            className="fd-button--light sap-icon--slim-arrow-down fd-select__button"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select /> create Select component 2`] = `
<div
  className="fd-popover"
>
  <div
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="fd-popover__control"
    >
      <div
        aria-controls="fd-ei5J6WhEr"
        aria-expanded={false}
        aria-haspopup={true}
        className="fd-select blue"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="fd-select__control"
        >
          <button
            className="fd-button--light sap-icon--slim-arrow-down fd-select__button"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select /> create Select component 3`] = `
<div
  className="fd-popover"
>
  <div
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="fd-popover__control"
    >
      <div
        aria-controls="fd-NKhEOTR-Q"
        aria-expanded={false}
        aria-haspopup={true}
        className="fd-select"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        standard={true}
        tabIndex={0}
      >
        <div
          className="fd-select__control"
        >
          <button
            className="fd-button--light sap-icon--slim-arrow-down fd-select__button"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select /> create Select component 4`] = `
<div
  className="fd-popover"
>
  <div
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="fd-popover__control"
    >
      <div
        aria-controls="fd-ZAwyY_V2Z"
        aria-expanded={false}
        aria-haspopup={true}
        className="fd-select"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="fd-select__control"
        >
          <button
            className="fd-button--light sap-icon--slim-arrow-down fd-select__button"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Select /> create Select component 5`] = `
<div
  className="fd-popover"
>
  <div
    onFocus={[Function]}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="fd-popover__control"
    >
      <div
        aria-controls="fd-Ijwr1gHMY"
        aria-expanded={false}
        aria-haspopup={true}
        className="fd-select"
        onClick={[Function]}
        onKeyPress={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="fd-select__control"
        >
          <button
            className="fd-button--light sap-icon--slim-arrow-down fd-select__button"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
